src/charm.py

```diff
@@ -5,19 +5,176 @@
 """Charmed operator for the 5G Webui service."""
 
 import logging
+from typing import Union
 
-from ops.charm import CharmBase
+from charms.data_platform_libs.v0.data_interfaces import (  # type: ignore[import]
+    DatabaseCreatedEvent,
+    DatabaseRequires,
+)
+from charms.observability_libs.v1.kubernetes_service_patch import (  # type: ignore[import]
+    KubernetesServicePatch,
+)
+from jinja2 import Environment, FileSystemLoader
+from lightkube.models.core_v1 import ServicePort
+from ops.charm import CharmBase, PebbleReadyEvent
 from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import Layer
 
 logger = logging.getLogger(__name__)
 
+BASE_CONFIG_PATH = "/etc/webui"
+CONFIG_FILE_NAME = "webuicfg.conf"
+DATABASE_RELATION_NAME = "database"
+DATABASE_NAME = "free5gc"
+
 
 class WebuiOperatorCharm(CharmBase):
     """Charm the service."""
 
     def __init__(self, *args):
         super().__init__(*args)
+        self._container_name = self._service_name = "webui"
+        self._container = self.unit.get_container(self._container_name)
+        self._database = DatabaseRequires(
+            self,
+            relation_name=DATABASE_RELATION_NAME,
+            database_name=DATABASE_NAME,
+            extra_user_roles="admin",
+        )
+        self.framework.observe(self.on.webui_pebble_ready, self._on_webui_pebble_ready)
+        self.framework.observe(self.on.database_relation_joined, self._on_webui_pebble_ready)
+        self.framework.observe(self._database.on.database_created, self._on_database_created)
+        self.framework.observe(self._database.on.endpoints_changed, self._on_database_created)
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            service_name="webui",
+            ports=[
+                ServicePort(name="urlport-http", port=5000),
+                ServicePort(name="grpc", port=9876),
+            ],
+        )
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Handle database created event."""
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+        if not self._container.exists(path=BASE_CONFIG_PATH):
+            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.defer()
+            return
+        config_file_content = self._render_config_file(
+            database_name=DATABASE_NAME, database_url=event.uris.split(",")[0]
+        )
+        if not self._config_file_is_written() or not self._config_file_content_matches(
+            content=config_file_content
+        ):
```

Comment:
I am not sure what we gain here instead of just writing out the file.

Reply:
I wanted to make sure we restart the service on the workload only if the content of the config file changed. I now realize even that objective wasn't properly implemented (the `restart` was missing). I removed it now and fixed the issue I had so on database created event, we will always set the pebble layer and restart the `webui` service. 

And that's same indication I gave on the `smf` and the `amf` PR's. We'll want to align on our approach here. FYI @saltiyazan @sanchezfdezjavier